
# RegressHaplo R Package

This package infers haplotypes from a BAM file using a penalized regression approach. To install the package:

....

## RegressHaplo Pipeline

The RegressHaplo analysis is split into components comprising the following pipeline:


**BAM file --> variant calls --> read\_table --> loci --> potential haplotypes --> parameters --> solutions --> haplotypes --> fasta file**

The function calls implementing each step of the pipeline are as follows. These calls should be executed sequentially.  Alternatively, the full pipeline can be executed using 
`full_pipeline(bam_file, out_dir, max_num_haplotypes=1200, rho_vals=c(.1,1,5,10), sig=.01, num_trials=100)`.   

Each component of the pipeline generates output files that are used by the following component.  Users can substitute their own files, thereby allowing a user to replace any given pipeline component with their own analysis.  For example, variant calling can be done with another tool.  See below for details regarding files generated by each component.

**The final haplotype reconstruction is contained in `final_haplo.fasta`.**  An example BAM file and accompanying RegressHaplo analysis and generated files are provided in the **`data/`** folder as a concrete example.

**Pipeline Components**:

1.  `bam_to_variant_calls.pipeline(bam_file, out_dir, start_pos=NULL, end_pos=NULL, sig=.01)` 
2. `variant_calls_to_read_table.pipeline(bam_file, out_dir)`
3. `read_table_to_loci.pipeline(out_dir, max_num_haplotypes=1200)`
4. `loci_to_haplotypes.pipeline(out_dir, max_num_haplotypes=1200)`
5. `haplotypes_to_parameters.pipeline(out_dir)`
6. `parameters_to_solutions.pipeline(out_dir, num_trials=100)`
7. `solutions_to_haplotypes.pipeline(out_dir)`
8. `haplotypes_to_fasta.pipeline(bam_file, out_dir)`

**Arguments**:

* `bam_file` - A BAM file of the NGS dataset.  The reference is not needed by RegressHaplo. 
* `out_dir` - directory in which output and input datafiles are written and read.
* `start_pos, end_pos` - positions on reference over which to make variant calls.  If NULL, then the whole reference is considered.
* `max_num_haplotypes` - maximum number of haplotypes to consider in a locus and globally.  The default of 1200 represents a ceiling, beyond that and the computations slow down considerably.  Lowering the maximum number of haplotypes will speed up the reconstruction but potentially reduce accuracy.
* `rho_values` - which rho values to use in constructing solutions for the penalized regression.  For each rho value, the regression is solved `num_trials` time using different starting points.
* `sig` - significance level at which to call variants.  A Bonferonni correction is automatically applied to the significance level accounting for multiplicity of positions.  **To turn off error correction, set sig=0**
* `num_trials` - number of times to solve the penalized regression in order to search for an optimal solution.  

**Files Generated by Pipeline Components**:

At each step in the pipeline, a file is produced in `out_dir`.  All input files are also expected to be in `out_dir`, except for the bam_file since its path is given as an argument in the first two steps of the pipeline (subsequent steps don't need the bam file).  For example, the function `bam_to_read_table.pipeline` expects `variant_calls.csv` to be in `out_dir` and writes `read_table.csv` to `out_dir`.

The pipeline format allows the user to edit the files generated at either step or to create their own files, thereby bypassing certain steps in the pipeline

1. `variant_calls.csv`  A file containing a data.frame with columns: pos, A, C, G, T, -.   pos gives the position number relative to the reference (which is not needed by RegressHaplo).  The A,C,G,T,- columns have entries of 0 or 1 specifying if a given nucleotide is a true variant (1) or error (0). 
2. `read_table.csv`  A file containing a data.frame with columns count, pos1, pos2, ..., posn.  posi is a number (e.g. the column pos1 will actually be names "56") giving the position of the ith variant call specified in `variant_calls.csv`.  Each row of the data.frame corresponds to a collection of reads that cover the same variable positions and have the same nucleotides at those positions....
3. `loci.csv` A file containing a data.frame with columns locus, pos.  The locus column gives the number of the locus, starting with 1 and incrementing by one.   The pos column gives the subset of variable positions (specified in `variant_calls.csv` contained in each locus.  The format of each entry is pos1+pos2+..+posn.  For example, 100+200+220 as an entry in the pos column represents a locus containing positions 100,200,220 relative to the reference.
4. `h.csv` Each row corresponds to a potential haplotype.   Each column has the position as a header.  Columns exactly correspond to variable positions.  
3. `y.csv, P.csv` The y vector and P matrix in the least squares \|y - Ph\|^2.
4. `solutions.csv`
5. `final_haplo.csv` - A csv file containing nucleotide values of reconstructed haplotypes, but only at the variable positions.  
6. `final_haplo.fasta` - A fasta file containing the reconstructed haplotypes over the whole reference.  Variable positions have nucleotide values given by `final_haplo.csv` and other positions are given the consensus value.  Each sequence in the fasta file has the form `haplotype#_$frequency` where `$frequency` is the haplotype frequency.



**Accessor Functions**:

For convenience, the following functions are defined to allow the user to access data in files produced by the RegressHaplo pipeline:

1. `get_variant_calls.pipeline(out_dir)` - returns the data.frame containing the variant calls; the data.frame is stored in `$out_dir/variant_calls.csv`
2. `get_read_table.pipeline(out_dir, raw=F)` - returns a data.frame containing either the raw read table (contained in `$out_dir/raw_read_table.csv`) or the read table produced after filtering out errors (contained in `$out_dir/read_table.csv`)
3. 'get_loci.pipeline(out_dir)' - returns a list of numeric vectors giving the positions, relative to the reference, composing each loci (accesses `$out_dir/loci.csv`)
4. `get_y.pipeline(out_dir), get_P.pipeline(out_dir)` - return a numeric vector, the y vector, and a numeric vector, the P matrix.  These functions access `y.csv` and `P.csv` respectively
5. `get_solutions_summary(out_dir)` - returns a list containig elemetns `df_stats`, `m_p`, and  `h`.  `df_stats` is a data.frame with columns rho, K, fit.  Each row of `df_stats` corresponds to a single solution of the penalized regression with rho set to the specified value.  K gives the number of haplotypes in the solutions and fit gives the least squared errors of the K haplotypes.  `m_p` is a numeric matrix.  Each row is the h value solved for in a single solution of the penalized regression.  `h` is a matrix giving the global haplotypes.  Each row is a haplotype and each column gives a position.   The number of columns in `h` is the number of columns in `m_p`.   The number of rows in `df_stats` is the number of rows in `m_p`.  
6. `get_fasta.pipeline(out_dir)` - returns a list containing the elements haplotypes and freq.  haplotypes is a character.vector, each entry of which gives a full haplotype covering the reference.  freq is a numeric.vector giving the inferred frequency of each haplotype.

